Support cross compilation of TFLite pip package

vinceab opened the pull request:
This pull request add the support of cross compiling variables.
It allow to cross compile TFLite library and python module for platforms other than aarch64, Rpi and iOS.
It has been successfully tested with Yocto Openembedded environment to build TensorFlow Lite for the new STM32MP1 Microprocessor target from STMicroelectronics.

It also remove some bashisms in pip build script that prevent to build TFLite in server machine with a dash shell. 

BR
Vincent

vinceab committed:
TFLite: pip package: support cross compilation environment variables

Add build environment variable to allow to cross compile TensorFlow Lite
pip package for other platform than Rpi or X86.

Signed-off-by: Vincent Abriou <vincent.abriou@st.com>

terryheo reviewed (commented):
Could you elaborate in which case bash is not available? Since all the scripts are based on bash, I'm not sure which script should avoid bashism.

vinceab commented:
Hi,

The build issue has been encounter in different build server on which the host machine has the dash shell set by default.
In that case, the execution of the lite/tools/pip_package/build_pip_package.sh script is failing with the following error:
build_pip_package.sh: 22:  Bad substitution

In this patch, I  am just focusing on the build_pip_package.sh because it is the one that is used from the Yocto environment build system to build the python package.

Vincent

terryheo commented:
The first line of build_pip_package.sh is "#!/usr/bin/env bash".
Can you install bash on the build server?

> Hi,
> 
> The build issue has been encounter in different build server on which the host machine has the dash shell set by default.
> In that case, the execution of the lite/tools/pip_package/build_pip_package.sh script is failing with the following error:
> build_pip_package.sh: 22: Bad substitution
> 
> In this patch, I am just focusing on the build_pip_package.sh because it is the one that is used from the Yocto environment build system to build the python package.
> 
> Vincent

vinceab commented:
I understand your point.
I will in that case abandoned the first patch and update the pull request.
Vincent

terryheo reviewed (dismissed):
LGTM

gbaned commented:
@vinceab Can you please address Ubuntu Sanity errors? Thanks!

vinceab commented:
@gbaned a new patch has been uploaded. that should fix Ubuntu Sanity errors.

terryheo commented on tensorflow/lite/tools/pip_package/build_pip_package.sh:
```diff
@@ -23,7 +23,11 @@ export TENSORFLOW_DIR="${SCRIPT_DIR}/../../../.."
 TENSORFLOW_VERSION=$(grep "_VERSION = " "${TENSORFLOW_DIR}/tensorflow/tools/pip_package/setup.py" | cut -d= -f2 | sed "s/[ '-]//g")
 export PACKAGE_VERSION="${TENSORFLOW_VERSION}${VERSION_SUFFIX}"
-BUILD_DIR="/tmp/tflite_pip/${PYTHON}"
+if [[ "${TENSORFLOW_CROSS_COMPILATION}" == "true" ]]; then
```
how to set TENSORFLOW_CROSS_COMPILATION?

vinceab replied:
Hi,
TENSORFLOW_CROSS_COMPILATION is a new variable introduce to not break the current way of building Tensorflow pip package.
It should be set by the compilation environment to warn the .sh script that the build directory should change to have a local build directory.

I'd rather like to not have this environment variable but in this case, the BUILD_DIR should be a local build directory for every one meaning BUILD_DIR="$'{SCRIPT_DIR}/gen/tflite_pip" 

What would you prefer?

terryheo replied:
What's the issue of using "/tmp/tflite_pip/${PYTHON}" for cross build?

BTW, we also need to update tensorflow/lite/tools/pip_package/Makefile to change it.

vinceab replied:
In some cross compilation environment (like Yocto OpenEmbedded), you don't have access the the root directory of your host computer to build. It is impossible to use /tmp/.

Yes your right about the Makefile if you decide to mode the build directory into a local directory.

Do you want me to update my patch taking into account only a local directory for the build?

terryheo replied:
I see. Yes, let's use local directory to build which is aligned with lite/tools/make/Makefile behavior. Could you apply it?

vinceab replied:
yes for sure, new patch will be available soon

vinceab replied:
New patch uploaded

vinceab committed:
TFLite: pip_package: use local build directory

In some cross compilation environment, access of the root directory of the
host is forbidden. To avoid any issue while compilation, the use of a local
directory is preferable.
Further it will be aligned with the lite/tools/make/Makefile behavior.

Signed-off-by: Vincent ABRIOU <vincent.abriou@st.com>

terryheo reviewed (approved):
Thanks for the update.
